nomad/node_endpoint.go

```diff
@@ -101,6 +102,52 @@ func (n *Node) Register(args *structs.NodeRegisterRequest, reply *structs.NodeUp
 
 	// Set the reply index
 	reply.Index = index
+
+	n.srv.peerLock.RLock()
+	defer n.srv.peerLock.RUnlock()
+	if err := n.updateNodeUpdateResponse(nil, reply); err != nil {
+		n.srv.logger.Printf("[ERR] nomad.client: failed to populate NodeUpdateResponse: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// updateNodeUpdateResponse assumes the n.srv.peerLock is held for reading.
+func (n *Node) updateNodeUpdateResponse(snap *state.StateSnapshot, reply *structs.NodeUpdateResponse) error {
+	reply.LeaderRPCAddr = n.srv.raft.Leader()
+
+	// Reply with config information required for future RPC requests
+	reply.Servers = make([]*structs.NodeServerInfo, 0, len(n.srv.localPeers))
+	for k, v := range n.srv.localPeers {
+		reply.Servers = append(reply.Servers,
+			&structs.NodeServerInfo{
+				RpcAdvertiseAddr: k,
+				RpcMajorVersion:  int32(v.MajorVersion),
+				RpcMinorVersion:  int32(v.MinorVersion),
+				Datacenter:       v.Datacenter,
+			})
+	}
+
+	// Capture all the nodes to obtain the node count
+	if snap == nil {
+		ss, err := n.srv.fsm.State().Snapshot()
```

Comment:
Remove this and get the snapshot in the register and add guards that the snapshot and reply are non-nil